Does not seem to work with GHC v9.2.4

While compiling a project using this nice plugin in a repository using GHC 9.2.4 I get the following error at build time:

```
qualified-imports-plugin> configure                                                
qualified-imports-plugin> Configuring qualified-imports-plugin-0.0.1...
qualified-imports-plugin> build         
qualified-imports-plugin> Preprocessing library for qualified-imports-plugin-0.0.1..
qualified-imports-plugin> Building library for qualified-imports-plugin-0.0.1..
qualified-imports-plugin> [1 of 1] Compiling QualifiedImportsPlugin
qualified-imports-plugin>               
qualified-imports-plugin> /tmp/stack-51368181e2a25a1c/qualified-imports-plugin-0.0.1/src/QualifiedImportsPlugin.hs:81:13: error:
qualified-imports-plugin>     • Couldn't match expected type: SDoc -> t
qualified-imports-plugin>                   with actual type: MsgEnvelope DecoratedSDoc
qualified-imports-plugin>     • The function ‘mkPlainWarnMsg’
qualified-imports-plugin>       is applied to three value arguments,
qualified-imports-plugin>         but its type ‘SrcSpan -> SDoc -> MsgEnvelope DecoratedSDoc’
qualified-imports-plugin>         has only two
qualified-imports-plugin>       In the expression:
qualified-imports-plugin>         mkPlainWarnMsg
qualified-imports-plugin>           (hsc_dflags env)
qualified-imports-plugin>           (UnhelpfulSpan $ unhelpfulOther "QualifiedImportsPlugin")
qualified-imports-plugin>           ("Unknown argument:" <+> text x)
qualified-imports-plugin>       In an equation for ‘msg’:
qualified-imports-plugin>           msg 
qualified-imports-plugin>             = mkPlainWarnMsg
qualified-imports-plugin>                 (hsc_dflags env)
qualified-imports-plugin>                 (UnhelpfulSpan $ unhelpfulOther "QualifiedImportsPlugin")
qualified-imports-plugin>                 ("Unknown argument:" <+> text x)
qualified-imports-plugin>     • Relevant bindings include
qualified-imports-plugin>         msg :: t (bound at src/QualifiedImportsPlugin.hs:80:11)
qualified-imports-plugin>    |          
qualified-imports-plugin> 81 |             mkPlainWarnMsg
qualified-imports-plugin>    |             ^^^^^^^^^^^^^^...
qualified-imports-plugin>               
qualified-imports-plugin> /tmp/stack-51368181e2a25a1c/qualified-imports-plugin-0.0.1/src/QualifiedImportsPlugin.hs:82:16: error:
qualified-imports-plugin>     • Couldn't match expected type ‘SrcSpan’
qualified-imports-plugin>                   with actual type ‘DynFlags’
qualified-imports-plugin>     • In the first argument of ‘mkPlainWarnMsg’, namely
qualified-imports-plugin>         ‘(hsc_dflags env)’
qualified-imports-plugin>       In the expression:
qualified-imports-plugin>         mkPlainWarnMsg
qualified-imports-plugin>           (hsc_dflags env)
qualified-imports-plugin>           (UnhelpfulSpan $ unhelpfulOther "QualifiedImportsPlugin")
qualified-imports-plugin>           ("Unknown argument:" <+> text x)
qualified-imports-plugin>       In an equation for ‘msg’:
qualified-imports-plugin>           msg 
qualified-imports-plugin>             = mkPlainWarnMsg
qualified-imports-plugin>                 (hsc_dflags env)
qualified-imports-plugin>                 (UnhelpfulSpan $ unhelpfulOther "QualifiedImportsPlugin")
qualified-imports-plugin>                 ("Unknown argument:" <+> text x)
qualified-imports-plugin>    |          
qualified-imports-plugin> 82 |               (hsc_dflags env)
qualified-imports-plugin>    |                ^^^^^^^^^^^^^^
qualified-imports-plugin>               
qualified-imports-plugin> /tmp/stack-51368181e2a25a1c/qualified-imports-plugin-0.0.1/src/QualifiedImportsPlugin.hs:83:16: error:
qualified-imports-plugin>     • Couldn't match expected type ‘SDoc’ with actual type ‘SrcSpan’
qualified-imports-plugin>     • In the second argument of ‘mkPlainWarnMsg’, namely
qualified-imports-plugin>         ‘(UnhelpfulSpan $ unhelpfulOther "QualifiedImportsPlugin")’
qualified-imports-plugin>       In the expression:
qualified-imports-plugin>         mkPlainWarnMsg
qualified-imports-plugin>           (hsc_dflags env)
qualified-imports-plugin>           (UnhelpfulSpan $ unhelpfulOther "QualifiedImportsPlugin")
qualified-imports-plugin>           ("Unknown argument:" <+> text x)
qualified-imports-plugin>       In an equation for ‘msg’:
qualified-imports-plugin>           msg 
qualified-imports-plugin>             = mkPlainWarnMsg
qualified-imports-plugin>                 (hsc_dflags env)
qualified-imports-plugin>                 (UnhelpfulSpan $ unhelpfulOther "QualifiedImportsPlugin")
qualified-imports-plugin>                 ("Unknown argument:" <+> text x)
qualified-imports-plugin>    |          
qualified-imports-plugin> 83 |               (UnhelpfulSpan $ unhelpfulOther "QualifiedImportsPlugin")
qualified-imports-plugin>    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
qualified-imports-plugin>               
qualified-imports-plugin> /tmp/stack-51368181e2a25a1c/qualified-imports-plugin-0.0.1/src/QualifiedImportsPlugin.hs:168:33: error:
qualified-imports-plugin>     • Couldn't match type ‘SrcSpan’
qualified-imports-plugin>                      with ‘SrcSpanAnn' (EpAnn AnnListItem)’
qualified-imports-plugin>       Expected: Maybe (XRec GhcPs ModuleName)
qualified-imports-plugin>         Actual: Maybe (Located ModuleName)
qualified-imports-plugin>     • In the ‘ideclAs’ field of a record
qualified-imports-plugin>       In the fourth argument of ‘maybe’, namely
qualified-imports-plugin>         ‘(simpleImportDecl n)
qualified-imports-plugin>            {ideclQualified = QualifiedPre, ideclAs = Just (noLoc qn),
qualified-imports-plugin>             ideclImplicit = True}’
qualified-imports-plugin>       In the expression:
qualified-imports-plugin>         (maybe noLoc L loc)
qualified-imports-plugin>           (simpleImportDecl n)
qualified-imports-plugin>             {ideclQualified = QualifiedPre, ideclAs = Just (noLoc qn),
qualified-imports-plugin>              ideclImplicit = True}
qualified-imports-plugin>     |         
qualified-imports-plugin> 168 |                       ideclAs = Just (noLoc qn),
qualified-imports-plugin>     |                                 ^^^^^^^^^^^^^^^
qualified-imports-plugin>               
qualified-imports-plugin> /tmp/stack-51368181e2a25a1c/qualified-imports-plugin-0.0.1/src/QualifiedImportsPlugin.hs:173:30: error:
qualified-imports-plugin>     • Couldn't match type ‘SrcSpan’
qualified-imports-plugin>                      with ‘SrcSpanAnn' (EpAnn AnnListItem)’
qualified-imports-plugin>       Expected: [LImportDecl GhcPs]
qualified-imports-plugin>         Actual: [Located (ImportDecl GhcPs)]
qualified-imports-plugin>     • In the ‘hsmodImports’ field of a record
qualified-imports-plugin>       In the second argument of ‘($)’, namely
qualified-imports-plugin>         ‘m {hsmodImports = hsmodImports m ++ newImports}’
qualified-imports-plugin>       In a stmt of a 'do' block:
qualified-imports-plugin>         return $ m {hsmodImports = hsmodImports m ++ newImports}
qualified-imports-plugin>     |         
qualified-imports-plugin> 173 |   return $ m {hsmodImports = hsmodImports m ++ newImports}
qualified-imports-plugin>     |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
qualified-imports-plugin>               
qualified-imports-plugin> /tmp/stack-51368181e2a25a1c/qualified-imports-plugin-0.0.1/src/QualifiedImportsPlugin.hs:178:12: error:
qualified-imports-plugin>     • Variable not in scope: unitState :: DynFlags -> b
qualified-imports-plugin>     • Perhaps you meant data constructor ‘UnitState’ (imported from GHC.Plugins)
qualified-imports-plugin>     |         
qualified-imports-plugin> 178 |   let us = unitState $ hsc_dflags env
qualified-imports-plugin>     |            ^^^^^^^^^
Progress 22/23                          
````

It seems like some internals of GHC that this plugin depends on might have changed.

--- a/ci.nix
+++ b/ci.nix
@@ -1,5 +1,5 @@
 let
-compilers = [ "ghc8104" "ghc901" ];
+compilers = [ "ghc8107Binary" "ghc902" "ghc924" ];
 withCompiler = c: (import ./default.nix {compiler=c;}).qualified-imports-plugin;
 in
 builtins.listToAttrs

--- a/default.nix
+++ b/default.nix
@@ -1,4 +1,4 @@
-{ compiler ? "ghc901" }:
+{ compiler ? "ghc92" }:
 
 let
   sources = import ./nix/sources.nix;
@@ -26,6 +26,7 @@ let
       pkgs.haskellPackages.cabal-install
       pkgs.haskellPackages.ghcid
       pkgs.haskellPackages.ormolu
+      pkgs.haskellPackages.haskell-language-server
       pkgs.niv
       pkgs.nixpkgs-fmt
     ];

--- a/example/example.cabal
+++ b/example/example.cabal
@@ -9,7 +9,7 @@ executable example
   ghc-options:      -fplugin=QualifiedImportsPlugin
   mixins:           base hiding (Prelude)
                   , relude (Relude as Prelude)
-  build-depends:    base ^>=4.15.0.0
+  build-depends:    base >=4.14 && <5.0
                   , relude
                   , text
                   , containers

--- a/nix/sources.json
+++ b/nix/sources.json
@@ -5,10 +5,10 @@
         "homepage": null,
         "owner": "nixos",
         "repo": "nixpkgs",
-        "rev": "800a3dd90970a277e3f6853633bd7faf04d6691e",
-        "sha256": "07gzvf3m3vh9kkig2x9iybv8yr04kvmrx665z0fb0n5h55ha4v8y",
+        "rev": "de80d1d04ee691279e1302a1128c082bbda3ab01",
+        "sha256": "1jyslvml9rlknix2w4h4rh7vs7vkqbvgq6d1z8y5mpa7sh5786av",
         "type": "tarball",
-        "url": "https://github.com/nixos/nixpkgs/archive/800a3dd90970a277e3f6853633bd7faf04d6691e.tar.gz",
+        "url": "https://github.com/nixos/nixpkgs/archive/de80d1d04ee691279e1302a1128c082bbda3ab01.tar.gz",
         "url_template": "https://github.com/<owner>/<repo>/archive/<rev>.tar.gz"
     }
 }

--- a/qualified-imports-plugin.cabal
+++ b/qualified-imports-plugin.cabal
@@ -16,7 +16,7 @@ library
   default-language:    Haskell2010
   exposed-modules:     QualifiedImportsPlugin
   ghc-options:         -Wall -fno-warn-name-shadowing
-  build-depends:       base >= 4.11 && < 5
+  build-depends:       base >=4.14 && <5.0
                      , ghc >= 8.10
                      , syb
                      , containers
@@ -30,6 +30,6 @@ test-suite tests
   default-language: Haskell2010
   ghc-options:      -Wall -threaded
                     -fplugin=QualifiedImportsPlugin
-  build-depends:    base >=4.11 && < 5
+  build-depends:    base
                   , qualified-imports-plugin
                   , text

--- a/src/QualifiedImportsPlugin.hs
+++ b/src/QualifiedImportsPlugin.hs
@@ -1,5 +1,4 @@
 {-# LANGUAGE CPP #-}
-{-# LANGUAGE LambdaCase #-}
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE TypeApplications #-}
 
@@ -79,7 +78,9 @@ parseOpts (x : xs) = case parseOpt x of
       -- 'WarnMsg -> Hsc ()'.
       let msg =
             mkPlainWarnMsg
+#if __GLASGOW_HASKELL__ < 902
               (hsc_dflags env)
+#endif
               (UnhelpfulSpan $ unhelpfulOther "QualifiedImportsPlugin")
               ("Unknown argument:" <+> text x)
        in return ((), consBag msg wm)
@@ -162,10 +163,10 @@ modifyHsMod opts m = do
             )
           & map
             ( \(loc, n, qn) ->
-                (maybe noLoc L loc)
+                (maybe noLocA noAnnLoc loc)
                   (simpleImportDecl n)
                     { ideclQualified = QualifiedPre,
-                      ideclAs = Just (noLoc qn),
+                      ideclAs = Just (noLocA qn),
                       -- This makes GHC not complain about unused imports.
                       ideclImplicit = True
                     }
@@ -175,7 +176,7 @@ modifyHsMod opts m = do
 -- Figure out if the module is available to import.
 isModuleAvailable :: HscEnv -> ModuleName -> Bool
 isModuleAvailable env n =
-  let us = unitState $ hsc_dflags env
+  let us = hsc_units env
    in case lookupModuleWithSuggestions us n Nothing of
         LookupFound _ _ -> True -- "found"
         LookupMultiple _ -> False -- "multiple"
@@ -203,3 +204,24 @@ referencedModules m =
               Just (L loc (Qual m _)) -> Map.singleton m loc
               Just _ -> mempty
         )
+
+-- Compat stuff
+
+#if __GLASGOW_HASKELL__ < 902
+noLocA :: a -> Located a
+noLocA = noLoc
+
+noAnnLoc :: SrcSpan -> a -> Located a
+noAnnLoc = L
+#else
+noAnnLoc :: SrcSpan -> e -> GenLocated (SrcAnn ann) e
+noAnnLoc = L . noAnnSrcSpan
+#endif
+
+#if __GLASGOW_HASKELL__ < 900
+hsc_units :: HscEnv -> DynFlags
+hsc_units = hsc_dflags
+#elif __GLASGOW_HASKELL__ < 902
+hsc_units :: HscEnv -> UnitState
+hsc_units = unitState . hsc_dflags
+#endif

